content/2020-05-22-book-review-writing-an-interpreter-in-go/index.md

```diff
@@ -0,0 +1,581 @@
+
+#### infix parse functions
+
+I wrote a note later where he says,
+
+> Yes, we need to register an `infixParseFn` for `token.LPAREN`. This way we parse the expression that is the function (either an identifier, or a function literal), then check for an `infixParseFn` associated with `token.LPAREN` and call it with the already parsed expression as argument. And in this `infixParseFn` we can then parse the argument list.
+>
+> -Ball 122
+
+I'm blocking myself here. In my mind, I think "infix === things like +, -, \*, /" so it feels strange here to use an LPAREN or "(" as an infix, but I guess infix can also mean in between two other characters. Example: `add(5, 5)` where the `(` is in between the "d" and the "5". Still, future me should revisit this.
+
+#### serialization
+
+Thorsten mentions this,
+
+> What we want is an AST that (serialized as a string)...
+>
+> -Ball 89
+
+And I thought, "Do I really know what this means?" And the answer is no, no I do not. Wikipedia [explains it](https://en.wikipedia.org/wiki/Serialization) well. I'll summarize it as taking something, translating it into a data structure or object and storing it for later use (possibly restructuring it).
+
+### Chapter 3 - Evaluation
+
+This is the part where the abstract syntax trees are turned into something meaningful.
+
+#### interpreters vs compilers
+
+> The notion of an interpreter as something that doesn't leave executable artifacts behind (in contract to a compiler, which does just that) gets fuzzy real fast when looking at the implementations of real-world and highly-optimized programming languages.
+>
+> -Ball 132
+
+I found this super helpful (at least comparing it to say Rust). Interpreter has one-step. Compilers have two.
+
+#### tree-walking interpreters
+
+I wrote down this because I wanted to look up a few examples of programming languages that fall into this category. I don't know how trustworthy [these Reddit comments](https://www.reddit.com/r/ProgrammingLanguages/comments/dfi9lk/viability_of_treewalk_interpreters/) are but here are a few notes:
+
+> Ruby was a tree-walk interpreter before version 1.9 (I think) and was used in production quite a bit, albeit with a reputation for being very slow and consuming a lot of RAM.
+
+> R was also a tree-walking interpreter for around 20 years! It only changed in the last 5 or so years, a similar time as Ruby.
+
+> It seems Perl still is a tree-walk interpreter.
+
+A bit later, Thorsten does touch on a few. He mentions the Ruby example.
+
+#### intermediate representation
+
+I hadn't heard this term before. According to [Wikipedia](https://en.wikipedia.org/wiki/Intermediate_representation),
+
+> intermediate representation is the [data structure](https://en.wikipedia.org/wiki/Data_structure) or code used internally by a [compiler](https://en.wikipedia.org/wiki/Compiler) or [virtual machine](https://en.wikipedia.org/wiki/Virtual_machine) to represent [source code](https://en.wikipedia.org/wiki/Source_code).
```

Comment:
FWIW, and I don't know if I mention this in the book or not, Intermediate Representation is often shortened to "IR". LLVM IR, for example, is the intermediate representation used by the LLVM compiler toolchain. That often pops up when there's talk around the Rust compiler, for example, which uses LLVM :)

Reply:
I'm sure you do! Ahhh, okay. That makes sense :) I heard LLVM, but it's never something I've understood. Thanks for mentioning that!